File: ALG/array/Find specific num count in Orderly array.m
//
//  Find specific num count in Orderly array.m
//  ALG
//
//  Created by junlongj on 2017/10/13.
//  Copyright © 2017年 Jincc. All rights reserved.
//

#import "Find specific num count in Orderly array.h"

@implementation Find_specific_num_count_in_Orderly_array

/**
 [1,2,3,4,4,4,4,6,8,9]

 @param array <#array description#>
 @param target <#target description#>
 @return <#return value description#>
 */
+(NSInteger)findSpecificCountInOrderlyArray:(NSArray *)array targetNum:(NSInteger)target{
    NSInteger left ,right;
    left = 0;
    right = array.count-1;
    NSInteger count = 0;
    while (left < right) {
        NSInteger mid = left + (right - left)/2;
        if ([array[mid] integerValue] == target) {
            //发现了target的位置,然后在它的位置周围找
            count +=1;
            NSInteger k = mid;
            NSInteger m = mid;
            while ([array[--k] integerValue] == target) {
                count+=1;
            }
            while ([array[++m] integerValue] == target) {
                count+=1;
            }
            return count;
        }else if ([array[mid] integerValue] < target){
            left = mid + 1;
        }else{
            right = mid - 1;
        }
    }
    return -1;
}

@end

File: C++/array/Find specific num count in Orderly array.cpp
//
//  main.cpp
//  LeetCode
//
//  Created by junl on 2019/1/21.
//  Copyright © 2019 junl. All rights reserved.
//

#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <cctype>
using namespace std;
/*
 在给定的一个已经排好序的数组中，找出指定数字出现的次数。例如数组[1,2,3,4,4,4,4,6,8,9]中4出现的次数为4次。
 */
int findSpecificCountInOrderlyArray(vector<int> &nums, int target){
    int count = 0;
    //首先通过二分查找找到对应的数
    int i = 0, j= nums.size() - 1, mid = i;
    while (i < j) {
        mid = i + (j - i) / 2;
        if (nums[mid] == target) {
            count++;
            int k = mid,m=mid;
            while (nums[--k] == target) {
                count++;
            }
            while (nums[++m] == target) {
                count++;
            }
            return count;
        }
        if (nums[mid] > target) {
            j = mid - 1;
        }else{
            i = mid + 1;
        }
    }
    return count;
};
int main(int argc, const char * argv[]) {
    vector<int> nums{1,2,3,4,4,4,4,6,8,9};
    cout << findSpecificCountInOrderlyArray(nums, 4) << endl;
    return 0;
}
